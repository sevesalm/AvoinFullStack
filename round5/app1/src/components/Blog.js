import React from "react";

const blogStyle = {
  padding: "0.5em",
  margin: "1em",
  border: "solid 1px #555",
  borderRadius: "5px"
};

const activeBlogStyle = {
  boxShadow: "2px 2px 8px #888"
};

const detailStyle = {
  margin: "0.5em 0 0 0.5em"
};

const BlogDetails = ({
  blog,
  blog: { id, url, likes, user },
  incrementLikes,
  loggedInUser,
  handleDeleteClick,
  askForConfirm
}) => (
  <div style={detailStyle}>
    <a href={url}>{url}</a>
    <div>
      {likes} likes <button onClick={incrementLikes(blog)}>Like</button>
    </div>
    <div>added by {user && user.name ? user.name : "Unknown"}</div>
    {(!user || user.username === loggedInUser.username) && (
      <button onClick={handleDeleteClick(id)}>
        {askForConfirm ? "Confirm?" : "Delete"}
      </button>
    )}
  </div>
);

class Blog extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showDetails: false,
      askForConfirm: false
    };
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  toggleVisibility = () => {
    this.setState({ showDetails: !this.state.showDetails });
  };

  handleDeleteClick = id => () => {
    if (this.state.askForConfirm) {
      this.props.deleteBlog(id);
    } else {
      this.setState({ askForConfirm: !this.state.askForConfirm });
      this.timer = setTimeout(
        () => this.setState({ askForConfirm: false }),
        2000
      );
    }
  };

  render() {
    const finalBlogStyle = this.state.showDetails
      ? { ...blogStyle, ...activeBlogStyle }
      : blogStyle;
    return (
      <div className="blog" style={finalBlogStyle}>
        <span className="blog-title" onClick={this.toggleVisibility}>
          {this.props.blog.title}
        </span>{" "}
        {this.props.blog.author}
        {this.state.showDetails && (
          <BlogDetails
            className="blog-details"
            blog={this.props.blog}
            incrementLikes={this.props.incrementLikes}
            loggedInUser={this.props.loggedInUser}
            handleDeleteClick={this.handleDeleteClick}
            askForConfirm={this.state.askForConfirm}
          />
        )}
      </div>
    );
  }
}

export default Blog;
