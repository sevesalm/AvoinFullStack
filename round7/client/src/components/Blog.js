import React from 'react';
import { connect } from 'react-redux';
import { blogActions } from '../reducers/blogReducer';

const blogStyle = {
  padding: '0.5em',
  margin: '1em',
  border: 'solid 1px #555',
  borderRadius: '5px'
};

const activeBlogStyle = {
  boxShadow: '2px 2px 8px #888'
};

const detailStyle = {
  margin: '0.5em 0 0 0.5em'
};

const BlogDetails = ({
  blog,
  blog: { id, url, likes, user },
  incrementLikes,
  loggedInUser,
  handleDeleteClick,
  askForConfirm
}) => (
  <div style={detailStyle}>
    <a href={url}>{url}</a>
    <div>
      {likes} likes <button onClick={() => incrementLikes(blog)}>Like</button>
    </div>
    <div>added by {user && user.name ? user.name : 'Unknown'}</div>
    {(!user || user.username === loggedInUser.username) && (
      <button onClick={handleDeleteClick(id)}>
        {askForConfirm ? 'Confirm?' : 'Delete'}
      </button>
    )}
  </div>
);

class Blog extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      askForConfirm: false
    };
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  handleDeleteClick = id => () => {
    if (this.state.askForConfirm) {
      this.props.deleteBlog(id);
    } else {
      this.setState({ askForConfirm: !this.state.askForConfirm });
      this.timer = setTimeout(
        () => this.setState({ askForConfirm: false }),
        2000
      );
    }
  };

  render() {
    const finalBlogStyle = { ...blogStyle, ...activeBlogStyle };
    return (
      <div className="blog" style={finalBlogStyle}>
        <span className="blog-title">{this.props.blog.title}</span>{' '}
        {this.props.blog.author}
        <BlogDetails
          className="blog-details"
          blog={this.props.blog}
          incrementLikes={this.props.incrementLikes}
          loggedInUser={this.props.loggedInUser}
          handleDeleteClick={this.handleDeleteClick}
          askForConfirm={this.state.askForConfirm}
        />
      </div>
    );
  }
}

export default connect(
  null,
  dispatch => ({
    dispatch,
    incrementLikes: blog => {
      const updatedBlog = {
        title: blog.title,
        author: blog.author,
        url: blog.url,
        likes: blog.likes + 1,
        user: blog.user ? blog.user._id : null
      };
      dispatch(blogActions.updateBlog(blog.id, updatedBlog));
    },
    deleteBlog: id => dispatch(blogActions.deleteBlog(id))
  })
)(Blog);
