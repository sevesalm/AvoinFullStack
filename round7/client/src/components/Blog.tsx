import React from 'react';
import { connect } from 'react-redux';
import { blogActions, IBlog } from '../reducers/blogReducer';
import {
  INotification,
  notificationActions
} from '../reducers/notificationReducer';

import { Button, Typography } from '@material-ui/core';
import { IUser } from '../reducers/userReducer';

const blogStyle = {
  border: 'solid 1px #555',
  borderRadius: '5px',
  margin: '1em',
  padding: '0.5em'
};

const activeBlogStyle = {
  boxShadow: '2px 2px 8px #888'
};

const BlogDetails = ({
  blog,
  incrementLikes,
  handleDeleteClick,
  login,
  askForConfirm
}: {
  blog: IBlog;
  incrementLikes: (blog: IBlog) => void;
  handleDeleteClick: (id: string) => (event: any) => void;
  login: IUser;
  askForConfirm: () => void;
}) => (
  <div className="blog-details">
    <Typography variant="body1">
      <a href={blog.url}>{blog.url}</a>
    </Typography>
    <Typography variant="body1">{blog.likes} likes</Typography>
    <Button
      color="primary"
      variant="contained"
      onClick={() => incrementLikes(blog)}
    >
      Like
    </Button>{' '}
    {(!blog.user || (login && blog.user.username === login.username)) && (
      <Button
        color="default"
        variant="contained"
        onClick={handleDeleteClick(blog.id)}
      >
        {askForConfirm ? 'Confirm?' : 'Delete'}
      </Button>
    )}
    <Typography variant="body1">
      added by {blog.user && blog.user.name ? blog.user.name : 'Unknown'}
    </Typography>
    {blog.comments.length ? (
      <div className="blog-comments">
        <Typography variant="body1">Comments</Typography>
        <ul>
          {blog.comments.map((item, idx) => (
            <li key={idx}>{item}</li>
          ))}
        </ul>
      </div>
    ) : (
      <Typography variant="body1">No comments</Typography>
    )}
  </div>
);

export class Blog extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
    this.state = {
      askForConfirm: false,
      comment: '',
      timer: null
    };
  }

  public componentWillUnmount() {
    clearTimeout(this.state.timer);
  }

  public handleDeleteClick = (id: string) => () => {
    if (this.state.askForConfirm) {
      this.props.deleteBlog(id, this.props.login.token);
    } else {
      this.setState({
        askForConfirm: !this.state.askForConfirm,
        timer: setTimeout(() => this.setState({ askForConfirm: false }), 2000)
      });
    }
  };

  public handleInputChange = (event: any) =>
    this.setState({ comment: event.target.value });

  public handleSubmit = async (event: any) => {
    event.preventDefault();
    this.props.commentBlog(this.props.blog.id, this.state.comment);
    this.setState({ comment: '' });
    this.props.showNotification({
      message: `Comment "${this.state.comment}" added to blog "${
        this.props.blog.title
      }"`,
      type: 'info'
    });
  };

  public render() {
    const finalBlogStyle = { ...blogStyle, ...activeBlogStyle };
    return (
      <div className="blog" style={finalBlogStyle}>
        <Typography variant="title" className="blog-title">
          {this.props.blog.title}
        </Typography>
        <Typography variant="body1" className="blog-author">
          by {this.props.blog.author}
        </Typography>
        <BlogDetails
          blog={this.props.blog}
          incrementLikes={this.props.incrementLikes}
          login={this.props.login}
          handleDeleteClick={this.handleDeleteClick}
          askForConfirm={this.state.askForConfirm}
        />
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            name="comment"
            value={this.state.comment}
            onChange={this.handleInputChange}
          />
          <Button type="submit">Add comment</Button>
        </form>
      </div>
    );
  }
}

export default connect(
  (state: any) => ({ login: state.login }),
  (dispatch: any) => ({
    commentBlog: (id: string, comment: string) =>
      dispatch(blogActions.commentBlog(id, comment)),
    deleteBlog: (id: string, token: string) =>
      dispatch(blogActions.deleteBlog(id, token)),
    dispatch,
    incrementLikes: (blog: IBlog) => {
      const updatedBlog = {
        author: blog.author,
        likes: blog.likes + 1,
        title: blog.title,
        url: blog.url,
        user: blog.user ? blog.user._id : null
      };
      dispatch(blogActions.updateBlog(blog.id, updatedBlog));
    },
    showNotification: (notification: INotification) =>
      dispatch(notificationActions.showNotification(notification))
  })
)(Blog);
