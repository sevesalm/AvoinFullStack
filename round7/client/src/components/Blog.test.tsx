import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import { Blog } from './Blog';

describe('Blog', () => {
  const testBlog = {
    author: 'Test Author',
    comments: ['Is good', 'Meh', 'Is baad'],
    likes: 10,
    title: 'Test Title',
    url: 'http://test.com'
  };

  it('shows blog details', () => {
    const blog = mount(<Blog blog={testBlog} />);
    const blogElement = blog.find('.blog').text();
    expect(blogElement).toContain(testBlog.title);
    expect(blogElement).toContain(testBlog.author);
    expect(blogElement).toContain(testBlog.url);
    const comments = blog.find('.blog-comments li');
    expect(comments.length).toEqual(3);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Blog blog={testBlog} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
