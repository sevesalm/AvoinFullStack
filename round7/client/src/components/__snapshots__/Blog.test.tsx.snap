// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog renders correctly 1`] = `
<div
  className="blog"
  style={
    Object {
      "border": "solid 1px #555",
      "borderRadius": "5px",
      "boxShadow": "2px 2px 8px #888",
      "margin": "1em",
      "padding": "0.5em",
    }
  }
>
  <h2
    className="MuiTypography-root-1 MuiTypography-title-7 blog-title"
  >
    Test Title
  </h2>
  <p
    className="MuiTypography-root-1 MuiTypography-body1-10 blog-author"
  >
    by 
    Test Author
  </p>
  <div
    className="blog-details"
  >
    <p
      className="MuiTypography-root-1 MuiTypography-body1-10"
    >
      <a
        href="http://test.com"
      >
        http://test.com
      </a>
    </p>
    <p
      className="MuiTypography-root-1 MuiTypography-body1-10"
    >
      10
       likes
    </p>
    <button
      className="MuiButtonBase-root-52 MuiButton-root-26 MuiButton-contained-37 MuiButton-containedPrimary-38 MuiButton-raised-40 MuiButton-raisedPrimary-41"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        Like
      </span>
      <span
        className="MuiTouchRipple-root-55"
      />
    </button>
     
    <button
      className="MuiButtonBase-root-52 MuiButton-root-26 MuiButton-contained-37 MuiButton-raised-40"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-27"
      >
        Delete
      </span>
      <span
        className="MuiTouchRipple-root-55"
      />
    </button>
    <p
      className="MuiTypography-root-1 MuiTypography-body1-10"
    >
      added by 
      Unknown
    </p>
    <div
      className="blog-comments"
    >
      <p
        className="MuiTypography-root-1 MuiTypography-body1-10"
      >
        Comments
      </p>
      <ul>
        <li>
          Is good
        </li>
        <li>
          Meh
        </li>
        <li>
          Is baad
        </li>
      </ul>
    </div>
  </div>
  <form
    onSubmit={[Function]}
  >
    <input
      name="comment"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="MuiButtonBase-root-52 MuiButton-root-26 MuiButton-text-28 MuiButton-flat-31"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-27"
      >
        Add comment
      </span>
      <span
        className="MuiTouchRipple-root-55"
      />
    </button>
  </form>
</div>
`;
